### PR TITLE
Guicing MetricsManager, CommonMetrics, and WebMetrics and handling related changes

### PR DESCRIPTION
**Pull request by reallocf**
Initial motivation was to remove ImmutableEnumChecker warnings. Worked with @kunkun-tang to determine that Guicing the offending classes was the best way to proceed.

MetricsManager, CommonMetrics, and WebMetrics had to be guicified together in order to handle tests that used both CommonMetrics/MetricsManager or WebMetrics/MetricsManager.

CommonMetricsTest and WebMetricsTest both required refactoring to handle the new guicier versions.

Added utility method to MetricsTestUtility that initializes a new version of MetricsManager that's used in multiple tests. This occurred in both CommonMetricsTest and WebMetricsTest as well as LoginAbstractAzkabanServletTest.

I'm hoping these testing changes/guicing overall will allow for more parallel testing. Additional adjustment of testing classes is likely necessary to make that happen.

Changes to all other files replaced the old singleton enum syntax with guice syntax.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java [hunk 1]
```diff
@@ -140,7 +140,7 @@ public void run() throws Exception {
         if (isMemGranted) {
           info(String.format("Memory granted for job %s", getId()));
           if (attempt > 1) {
-            CommonMetrics.INSTANCE.decrementOOMJobWaitCount();
+            SERVICE_PROVIDER.getInstance(CommonMetrics.class).decrementOOMJobWaitCount();
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: Not sure if this is a good practice. If me, I would have a CommonMetrics's instance in the Constructor of `ProcessJob`. Basically, I don't know the performance/efficiency of using `SERVICE_PROVIDER.getInstance`

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 2]
```diff
@@ -16,26 +16,29 @@
...
   private final AtomicLong dbConnectionTime = new AtomicLong(0L);
   private final AtomicLong OOMWaitingJobCount = new AtomicLong(0L);
   private final MetricRegistry registry;
   private Meter dbConnectionMeter;
   private Meter flowFailMeter;
 
+  @Inject
   CommonMetrics() {
-    this.registry = MetricsManager.INSTANCE.getRegistry();
+    this.registry = SERVICE_PROVIDER.getInstance(MetricsManager.class).getRegistry();
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: Sort of believe We should have MetricManager in parameter given this case.

File: azkaban-exec-server/src/main/java/azkaban/execapp/ExecMetrics.java [hunk 3]
```diff
@@ -29,7 +31,7 @@
   private final MetricRegistry registry;
 
   ExecMetrics() {
-    this.registry = MetricsManager.INSTANCE.getRegistry();
+    this.registry = SERVICE_PROVIDER.getInstance(MetricsManager.class).getRegistry();
```
Comment (kunkun-tang) [inline azkaban-exec-server/src/main/java/azkaban/execapp/ExecMetrics.java]: Sort of believe We should have MetricManager in parameter given this case.

File: azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java [hunk 4]
```diff
@@ -34,7 +36,7 @@ public AbstractJdbcLoader(final Props props) {
 
   protected Connection getDBConnection(final boolean autoCommit) throws IOException {
     Connection connection = null;
-    CommonMetrics.INSTANCE.markDBConnection();
+    SERVICE_PROVIDER.getInstance(CommonMetrics.class).markDBConnection();
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java]: Consider saving this reference in the constructor? This class can be considered a dependency of this class.

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java (hunk 2 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: Why not take MetricRegistry directly as a dependency and have Guice create it?

File: azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java [hunk 5]
```diff
@@ -28,15 +28,15 @@
...
  * The singleton class, MetricsManager, is the place to have MetricRegistry and ConsoleReporter in
  * this class. Also, web servers and executors can call {@link #startReporting(String, Props)} to
  * start reporting AZ metrics to remote metrics server.
  */
-public enum MetricsManager {
-  INSTANCE;
+public class MetricsManager {
 
   private static final Logger logger = Logger.getLogger(MetricsManager.class);
   private final MetricRegistry registry = new MetricRegistry();
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java]: Declare MetricRegistry as a dependency and have Guice create it for you?

File: azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java [hunk 6]
```diff
@@ -16,41 +16,39 @@
...
+    testUtil = new MetricsTestUtility(
+        SERVICE_PROVIDER.getInstance(MetricsManager.class).getRegistry());
+    metrics = SERVICE_PROVIDER.getInstance(CommonMetrics.class);
   }
 
   @Test
   public void testDBConnectionTimeMetrics() {
-    this.metrics.setDBConnectionTime(14);
-    assertEquals(14, this.testUtil.getGaugeValue("dbConnectionTime"));
+    metrics.setDBConnectionTime(14);
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java]: Did you apply saveaction?

Comment (reallocf) [inline azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java]: I did - metrics is static now because JUnit requires setUp to be static in order for it to be annotated @ BeforeClass which causes metrics and testUtil to be static. So "this." is no longer recommended by intelliJ. Does that make sense?

File: azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java [hunk 7]
```diff
@@ -16,41 +16,39 @@
...
-    // todo HappyRay: move MetricsManager, CommonMetrics to use Juice.
-    this.testUtil = new MetricsTestUtility(MetricsManager.INSTANCE.getRegistry());
-    this.metrics = CommonMetrics.INSTANCE;
+  private static MetricsTestUtility testUtil;
+  private static CommonMetrics metrics;
+
+  @BeforeClass
+  public static void setUp() {
+    // initialize new guice MetricsManager
+    MetricsTestUtility.initServiceProvider();
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java]: Here is an alternative:

CommonMetrics constructor takes an instance of MetricRegistry. In the production code, it is constructed using Guice.

In test code, simply new the instances you need without using Guice.

File: azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java (hunk 6 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java]: I see. Why do they need to be in BeforeClass?

Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java]: Typically it is better to have tests to be independent of each other as possible i.e. create their own instances when needed.

File: azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java (hunk 5 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java]: +1.

File: azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java [hunk 8]
```diff
@@ -16,41 +16,39 @@
...
+    testUtil = new MetricsTestUtility(
+        SERVICE_PROVIDER.getInstance(MetricsManager.class).getRegistry());
+    metrics = SERVICE_PROVIDER.getInstance(CommonMetrics.class);
```
Comment (kunkun-tang) [inline azkaban-common/src/test/java/azkaban/metrics/CommonMetricsTest.java]: We probably need to move this statement (`metrics = SERV...`) into separate tests, since we are testing itself specifically.

File: azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java [hunk 9]
```diff
@@ -16,25 +16,31 @@
 
 package azkaban.database;
 
+import static azkaban.ServiceProvider.SERVICE_PROVIDER;
+
 import azkaban.metrics.CommonMetrics;
 import azkaban.utils.Props;
 import java.io.IOException;
 import java.sql.Connection;
+import javax.inject.Inject;
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java]: What is this used for?

File: azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java [hunk 10]
```diff
@@ -16,25 +16,31 @@
...
 
 public abstract class AbstractJdbcLoader {
 
   private final AzkabanDataSource dataSource;
+  private final CommonMetrics commonMetrics;
 
-  public AbstractJdbcLoader(final Props props) {
+  @Inject
+  public AbstractJdbcLoader(final Props props, final CommonMetrics commonMetrics) {
     this.dataSource = DataSourceUtils.getDataSource(props);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java]: Not in this PR, but in the future, we should get datasource from Juice. 
@kunkun-tang

File: azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java [hunk 11]
```diff
@@ -43,7 +49,8 @@ protected Connection getDBConnection(final boolean autoCommit) throws IOExceptio
       DbUtils.closeQuietly(connection);
       throw new IOException("Error getting DB connection.", e);
     }
-    CommonMetrics.INSTANCE.setDBConnectionTime(System.currentTimeMillis() - startMs);
+    SERVICE_PROVIDER.getInstance(CommonMetrics.class)
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/AbstractJdbcLoader.java]: Shouldn't this use this.commonMetrics too?

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: +1

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 12]
```diff
@@ -16,26 +16,29 @@
...
   private final AtomicLong dbConnectionTime = new AtomicLong(0L);
   private final AtomicLong OOMWaitingJobCount = new AtomicLong(0L);
   private final MetricRegistry registry;
   private Meter dbConnectionMeter;
   private Meter flowFailMeter;
 
-  CommonMetrics() {
-    this.registry = MetricsManager.INSTANCE.getRegistry();
+  @Inject
+  public CommonMetrics(MetricsManager metricsManager) {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: Why not depend on MetircsRegistry directly?

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java [hunk 13]
```diff
@@ -66,7 +69,8 @@ private ExecutorManager createMultiExecutorManagerInstance(
 
     loader.addExecutor("localhost", 12345);
     loader.addExecutor("localhost", 12346);
-    return new ExecutorManager(props, loader, new AlerterHolder(props));
+    return new ExecutorManager(props, loader, new AlerterHolder(props),
+        new CommonMetrics(new MetricsManager(new MetricRegistry())));
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: Will a mock work here?

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java [hunk 14]
```diff
@@ -705,7 +705,8 @@ private static Props loadConfigurationFromAzkabanHome() {
 
   private void startWebMetrics() throws Exception {
 
-    final MetricRegistry registry = MetricsManager.INSTANCE.getRegistry();
+    final MetricRegistry registry = SERVICE_PROVIDER.getInstance(MetricsManager.class)
```
Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: Why not just get the MetricsRegistry directly?

Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: Have you tried binding MetricsRegistry class to a singleton in a GuiceModule?

Comment (reallocf) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: This is being refactored to rely on MetricsManager dependency which is handled in constructor. The MetricsManager class is used explicitly on line 736 to take advantage of the startReporting method. Do you think it'd be better to depend on both MetricsManager and MetricsRegistry?

Also, MetricsRegistry is binded to a singleton locally, I'm resolving some consequences of that switch then I'll push a new version - thanks for the feedback! :+1:

Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: I see. I would suggest listing all the direct dependencies directly to make the dependencies more explicit.

Comment (reallocf) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: That's a good point. I'll have it directly depend on both MetricsManager and MetricsRegistry then. Thanks again!

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java [hunk 15]
```diff
@@ -158,4 +163,10 @@ public FileSystem createHadoopFileSystem(final Configuration hadoopConf) {
   public QueryRunner createQueryRunner(final AzkabanDataSource dataSource) {
     return new QueryRunner(dataSource);
   }
+
+  @Provides
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: Why not 
    bind(MetricRegistry.class).in(Scopes.SINGLETON);
it also seems to work.

File: azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java [hunk 16]
```diff
@@ -28,24 +28,26 @@
 import java.lang.reflect.Constructor;
 import java.time.Duration;
 import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java]: Is this right? 
Not com.google.inject.Inject;?

File: azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java [hunk 17]
```diff
@@ -53,6 +59,20 @@ public JobTypeManagerTest() {
 
   @Before
   public void setUp() throws Exception {
+    // TODO: reallocf Remove injector creation/deletion when ProcessJob is fully guicified
+    final Props props = new Props();
+
+    final Injector injector = Guice.createInjector(new AbstractModule() {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java]: Why is this change necessary?

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecServerModule.java [hunk 18]
```diff
@@ -36,6 +36,6 @@ protected void configure() {
     bind(AzkabanExecutorServer.class).in(Scopes.SINGLETON);
     bind(TriggerManager.class).in(Scopes.SINGLETON);
     bind(FlowRunnerManager.class).in(Scopes.SINGLETON);
-
+    // TODO: add ExecMetrics when guicy
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecServerModule.java]: Is this Todo necessary?

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java [hunk 19]
```diff
@@ -172,6 +174,8 @@ public AzkabanWebServer(final Server server, final Props props) throws Exception
     this.executorManager = SERVICE_PROVIDER.getInstance(ExecutorManager.class);
     this.projectManager = SERVICE_PROVIDER.getInstance(ProjectManager.class);
     this.triggerManager = SERVICE_PROVIDER.getInstance(TriggerManager.class);
+    this.metricsManager = SERVICE_PROVIDER.getInstance(MetricsManager.class);
+    this.registry = new MetricRegistry();
```
Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: Shouldn't get it from Serivce_provider too?

File: azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java [hunk 20]
```diff
@@ -38,8 +37,13 @@
   // How long does user log fetch take when user call fetch-log api.
   private final AtomicLong logFetchLatency = new AtomicLong(0L);
 
-  WebMetrics() {
-    this.registry = MetricsManager.INSTANCE.getRegistry();
+  @Inject
+  WebMetrics(final MetricRegistry registry) {
+    // Add unit test coverage to guarantee no duplication
+    // Seeing this error after guicifying, but before binding as singleton in AzkabanWebServerModule:
```
Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java]: Could you find out what the problem is and then remove this comment?

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java [hunk 21]
```diff
@@ -17,22 +17,46 @@
...
 import javax.servlet.http.HttpServletResponse;
+import org.junit.Before;
 import org.junit.Test;
 
 public class LoginAbstractAzkabanServletTest {
 
+  @Before
+  public void setUp() {
+    // TODO: reallocf remove setUp() when LoginAbstractServlet fully guicified
+    final Props props = new Props();
```
Comment (HappyRay) [inline azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java]: If this logic is need, please find a way to reduce the code duplication.

see azkaban.test.Utils#initServiceProvider

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java (hunk 15 above)
Comment (reallocf) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: This was a recommendation @kunkun-tang made because he said the usage was similar to QueryRunner

Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: @kunkun-tang could you explain?

File: azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java (hunk 16 above)
Comment (reallocf) [inline azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java]: Nice catch! Thank you!

Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java]: In general, try to make small incremental changes. When a change becomes bigger, it is more likely to miss issues. It takes longer to finish reviews too.

Comment (reallocf) [inline azkaban-common/src/main/java/azkaban/metrics/MetricsManager.java]: Definitely, I appreciate the feedback. This one kind of got away from me as I was dealing with unraveling the dependencies. I also didn't understand Guice very well going into it, so it was a lot of learning on the fly 😅

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java (hunk 19 above)
Comment (reallocf) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: This is a consequence of the change that @kunkun-tang suggested in AzkabanCommonModule.

File: azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java (hunk 20 above)
Comment (reallocf) [inline azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java]: Yes, the problem has already been fixed. I just included this because there was no unit test confirming that WebMetrics is a singleton, which allowed me to build and run and then find the bug while using my local version. Adding `bind(WebMetrics.class).in(Scopes.SINGLETON);` to AzkabanWebServerModule fixed the problem, but it would probably be good to have a unit test in case anybody removes it for whatever reason. I didn't think it'd be wise to include it in this (already very large) PR, so I added the comment. What do you think?

File: azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java (hunk 17 above)
Comment (reallocf) [inline azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java]: This was included in order to handle the partial guicing of the ProcessJob class where I leverage the SERVICE_PROVIDER instead of injecting it the right way.

The problem with doing it the right was is then I have to refactor every class that extends the ProcessJob and then refactor every test for the classes that extend the ProcessJob and then suddenly I'm refactoring another 10+ files. I'd rather break this up into separate PR's which is why I'm leveraging this bandaid for now. Do you think it'd be better to implement Guice throughout this dependency chain now?

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java (hunk 21 above)
Comment (reallocf) [inline azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java]: I can do that. Do you think it's a good idea to build that infrastructure when this code will (hopefully) be retired quickly?

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java [hunk 22]
```diff
@@ -99,10 +99,11 @@
   @Inject
   public AzkabanExecutorServer(final Props props,
       final ExecutorLoader executionLoader,
-      final FlowRunnerManager runnerManager) throws Exception {
+      final FlowRunnerManager runnerManager, final MetricsManager metricsManager) throws Exception {
```
Comment (juhoautio) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: Would it be possible to not touch `AzkabanExecutorServer` yet? This earlier PR has conflicting changes: https://github.com/azkaban/azkaban/pull/1113. Or, if you're going even further with injection on this class, maybe you want to check out that PR so that you don't have to figure out the same things again.

File: azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java (hunk 20 above)
Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java]: Sounds good. Thanks for thinking about how to prevent regressions.

Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/WebMetrics.java]: IMHO, I prefer annotating the class itself with singleton if the class is designed to be used as singleton everywhere. It's less code and the information is centralized. Otherwise, readers have to check the module files too to find this information.

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java (hunk 21 above)
Comment (HappyRay) [inline azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java]: Yes if it is not a lot of work.
Can you reuse the code I referenced?  see azkaban.test.Utils#initServiceProvider

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java (hunk 22 above)
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: I think this is a different change. I don't expect resolving conflict will be too hard.

File: azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java (hunk 17 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java]: Yes, I am in favor of phased approach.

In this case, I replaced your Guice initiailization code with 
import static azkaban.test.Utils.initServiceProvider;

initServiceProvider();

And the tests in this file passed

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java (hunk 22 above)
Comment (juhoautio) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: True.

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java (hunk 19 above)
Comment (kunkun-tang) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: I just replaced this line by `this.registry = SERVICE_PROVIDER.getInstance(MetricRegistry.class);` and didn't observe any issues. What did you see if you use `Serivce_provider ` ?

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java (hunk 15 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: I was trying to google the difference between using `@provide` and `to bind`, but failed. Some online resources say that there'w no difference between the two approaches. I believe @suvodeep-pyne has better understanding on this case.

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java [hunk 23]
```diff
@@ -162,7 +170,8 @@ public void testSetupExecutorsException() throws ExecutorManagerException {
     Executor executor1 = loader.addExecutor("localhost", 12345);
 
     ExecutorManager manager =
-      new ExecutorManager(props, loader, new AlerterHolder(props));
+      new ExecutorManager(props, loader, new AlerterHolder(props),
```
Comment (kunkun-tang) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: Is there some way that we could consolidate the code that set up an Executor Manager.

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java (hunk 15 above)
Comment (reallocf) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: Should we just use to bind then in order to standardize on that?

File: azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java (hunk 17 above)
Comment (reallocf) [inline azkaban-common/src/test/java/azkaban/jobtype/JobTypeManagerTest.java]: Awesome, thank you for doing that! I'll make the update 👍

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java (hunk 21 above)
Comment (reallocf) [inline azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java]: Will do!

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java (hunk 23 above)
Comment (reallocf) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: :+1:

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java [hunk 24]
```diff
@@ -83,6 +86,9 @@ protected void configure() {
     bind(DataSource.class).to(AzkabanDataSource.class);
     bind(ExecutorManager.class).in(Scopes.SINGLETON);
     bind(AlerterHolder.class).in(Scopes.SINGLETON);
+    bind(CommonMetrics.class).in(Scopes.SINGLETON);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: Since CommonMetrics is annotated as singleton already, I think you don't need this line here. same with MetricsManager.

File: azkaban-exec-server/src/main/java/azkaban/execapp/ExecMetrics.java [hunk 25]
```diff
@@ -29,7 +31,8 @@
   private final MetricRegistry registry;
 
   ExecMetrics() {
-    this.registry = MetricsManager.INSTANCE.getRegistry();
+    // TODO: reallocf make guicy
+    this.registry = SERVICE_PROVIDER.getInstance(MetricsManager.class).getRegistry();
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/ExecMetrics.java]: Why not directly get the MetricsRegistry instance?

File: azkaban-exec-server/src/test/java/azkaban/execapp/AzkabanExecutorServerTest.java [hunk 26]
```diff
@@ -86,7 +87,9 @@ public void testInjection() throws Exception {
         new AzkabanCommonModule(props),
         new AzkabanExecServerModule()
     );
+    SERVICE_PROVIDER.unsetInjector();
```
Comment (HappyRay) [inline azkaban-exec-server/src/test/java/azkaban/execapp/AzkabanExecutorServerTest.java]: Why is this change necessary?

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServerModule.java [hunk 27]
```diff
@@ -32,5 +32,6 @@
   protected void configure() {
     bind(Server.class).toProvider(WebServerProvider.class);
     bind(AzkabanWebServer.class).in(Scopes.SINGLETON);
+    bind(WebMetrics.class).in(Scopes.SINGLETON);
```
Comment (HappyRay) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServerModule.java]: Same as the previous comment.

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java (hunk 24 above)
Comment (reallocf) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: Would it be a good idea to standardize on the @Singleton annotation? Not in this PR, but in the future.

File: azkaban-exec-server/src/test/java/azkaban/execapp/AzkabanExecutorServerTest.java (hunk 26 above)
Comment (reallocf) [inline azkaban-exec-server/src/test/java/azkaban/execapp/AzkabanExecutorServerTest.java]: Looks like it isn't anymore! It was necessary when things were not guiced the right way, but it seems to be working now without the change. Thanks!!

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java (hunk 23 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: This is an example of why Guice is helpful. Otherwise, we may have more similar code in the production code.
